### DISCUSSION TITLE
NATS Core - Custom Inbox Prefix on a per-message basis?

### DISCUSSION THREAD

**Discussion by VeryStrongFingers** [nats-io/nats.go#1740] [Ideas]
tl;dr: is it feasible to have something akin to:
```
nc, _ := nats.Connect("nats://127.0.0.1:4222")
nc.Request("eggs", nil, 5* time.Second, nats.CustomInboxPrefix("_YEETBOX"))
```

I'm trying to determine whether NATS (Core) Request-Reply "Inbox Prefix" functionality is possible to use on a per-message basis, with my goal being dynamically being able to set the message prefix on a per-request basis.

Based on `nats-cli` (`--inbox-prefix`) and `nats.go` (`nats.CustomInboxPrefix()`) the custom inbox prefix is coupled on the NATS Connection itself.
Obviously there are some work-arounds (hacks) I could attempt to workaround this limitation, but I thought it made more sense to ask if this is an intentional protocol design decision? or whether it could be justified to move the custom message prefix logic against messages rather than the connection.

References:
- https://natsbyexample.com/examples/auth/private-inbox-jwt/cli
- https://github.com/nats-io/nats.go/issues/1262
```go
// Golang
nc, err := nats.Connect(natsUrl, nats.CustomInboxPrefix("_YEETBOX"))
```
```bash
# CLI
nats --context joe --inbox-prefix _INBOX_joe request 'services.greet' ''
```

**Comment by wallyqs** [comment]
How custom do you want it to be? Could you share some examples of the usage you are looking for? You can implement request/response in NATS fairly easily and customize the inbox as you want: https://github.com/nats-io/nats.go/blob/main/nats.go#L4139-L4156

That approach (old style approach) is heavy in the subscription interest graph though, the built-in nats request/response handler tries to mitigate that by having a single subscription muxing all the responses on a single one looking like this by default: `_INBOX.<unique_string>.*`.  The support that exists right now is just to change that `_INBOX` prefix to a custom one (since there is nothing in the nats protocol defined related to how the string from an inbox should look like).
